### PR TITLE
Make ReadOnlyPerson observable #292

### LINKED ISSUE #292: Make ReadOnlyPerson observable
Currently we use ugly hacks all over the place in order to ensure that the UI is updated when we modify a Person object.

e.g. https://github.com/se-edu/addressbook-level4/blob/f40ee42cb2e29a99c9c6b8757b64b0b0f4284d67/src/main/java/seedu/address/model/person/UniquePersonList.java#L61-L63 
```java
        // TODO: The code below is just a workaround to notify observers of the updated person.
        // The right way is to implement observable properties in the Person class.
        // Then, PersonCard should then bind its text labels to those observable properties.
```

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
Fixes #292. Note that tags are still not completely binded and we still have to manually initialize it, though it now updates automatically whenever there are changes made to the tags.

```
ReadOnlyPerson is not observable. Whenever a ReadOnlyPerson is added 
or updated, PersonListPanel creates a new PersonCard and initialises 
it with the ReadOnlyPerson's details.

This is inefficient as multiple cards have to be deleted and
re-created whenever a ReadOnlyPerson is updated. Also, there is an
additional call to set the ReadOnlyPerson in the list of PersonCards
to update the view i.e updating the ReadOnlyPerson does not 
automatically update the view.

Let's update ReadOnlyPerson to be observable and teach PersonCard
to bind its view to the ReadOnlyPerson objects so that whenever they
are updated, the view is automatically updated. As such, the cards
do not need to be deleted and re-created.
```

### CODE REVIEW

**Comment by Zhiyuan-Amos** [comment]
@yamgent halp

**Comment by yamgent** [comment]
You might want to look at this regarding the `Tag`s. I was still hoping that we do not need to manually create the Tags on our own, because then it is not truly fully tied otherwise.

https://github.com/se-edu/addressbook-level4/pull/329#issuecomment-285670433

**Comment by Zhiyuan-Amos** [comment]
Yup, but it seems like it's at the expense of a lot more code. I.e you need the `MappedList` etc. I'm thinking that for a first improvement, this seems not too bad.

**Comment by Zhiyuan-Amos** [comment]
@damithc, while working on this PR, I noticed that `Logic` returns an `ObservableList` instead of an `UnmodifiableObservableList`.

https://github.com/se-edu/addressbook-level4/blob/master/src/main/java/seedu/address/logic/Logic.java#L23

Is there a rationale for that? It seems weird because `ModelManager` returns an `UnmodifiableObservableList` to `Logic`, but `Logic` returns an `ObservableList` to its callers (only `MainWindow` iirc). `MainWindow` then passes this list to init `PersonListPanel`, and `PersonListPanel` should not update the list as well.

**Comment by damithc** [comment]
> Is there a rationale for that?

Not that I know of.

**Comment by CanIHasReview** [comment]
# v1

@Zhiyuan-Amos submitted v1 for review.

* [v1 1/3] [Teach Person and ReadOnlyPerson to be observable](https://github.com/se-edu/addressbook-level4/pull/552/commits/8de20fb48300e84961de317a061f20cf274e9a8c)
* [v1 2/3] [PersonListPanel: Update ListView<ReadOnlyPerson> to ListView<PersonCard>](https://github.com/se-edu/addressbook-level4/pull/552/commits/72df9b06c25a40627a790168a90841e49b9f11bb)
* [v1 3/3] [UniquePersonList: Remove unnecessary method call](https://github.com/se-edu/addressbook-level4/pull/552/commits/9a0cb05638dc6ad1072fc3de740f878ae129872b)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/3e71299fdd455bdd42ba4768bcc26fbddbd82cfc...9a0cb05638dc6ad1072fc3de740f878ae129872b))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/552/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
> Not that I know of.

Noted, I have opened an issue for this.

**Comment by Zhiyuan-Amos** [comment]
@pyokagan, previously @MightyCupcakes tried resolving this issue (#329), and you responded that you wanted to rethink about this issue (https://github.com/se-edu/addressbook-level4/pull/329#issuecomment-285952945)

Would you like to take a look at my PR? Thanks!

File: src/main/java/seedu/address/ui/PersonListPanel.java [hunk 1]
```diff
@@ -21,20 +23,39 @@
...
+        initCardList(personList);
         setConnections(personList);
         registerAsAnEventHandler(this);
     }
 
+    private void initCardList(ObservableList<ReadOnlyPerson> personList) {
+        cardList = FXCollections.observableArrayList();
+        for (int i = 1; i <= personList.size(); i++) {
+            cardList.add(new PersonCard(personList.get(i - 1), i));
+        }
```
Comment (yamgent) [inline src/main/java/seedu/address/ui/PersonListPanel.java]: Seems like a repeat of `setListenerForListChange(...)`'s listener lambda's logic (btw any reason why this one starts with `i =1` but the other one starts with `i=0`?).

Is `ListChange` not invoked when the ListView is first created? (i.e. We are doing the same thing twice at startup?)

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/PersonListPanel.java]: Yup just noticed that it looks very similar. This method is meant to init the card list (I think the other way you suggested is to do the mapping thing, which perhaps was the way Jeremy did previously). I think ideally there isn't a need to init this list but hmm I'm not too sure.

Comment (yamgent) [inline src/main/java/seedu/address/ui/PersonListPanel.java]: Hmm okay then we can keep it here (but possible to refactor the method out?)

**Comment by Zhiyuan-Amos** [comment]
@pyokagan @yamgent I'm wondering if we can use EasyBind (https://github.com/TomasMikula/EasyBind#static-methods).

**Comment by CanIHasReview** [comment]
# v2

@Zhiyuan-Amos submitted v2 for review.

* [v2 1/4] [Teach Person and ReadOnlyPerson to be observable](https://github.com/se-edu/addressbook-level4/pull/552/commits/9c39fc83e40bfefa7e0920ec01ffbdd96415a862)
* [v2 2/4] [Import EasyBind](https://github.com/se-edu/addressbook-level4/pull/552/commits/0ba5a19ba8e89b33af0cc4fdd32505cba7eda0ba)
* [v2 3/4] [PersonListPanel: Update ListView<ReadOnlyPerson> to ListView<PersonCard>](https://github.com/se-edu/addressbook-level4/pull/552/commits/7d500e30efd7c420aed7b38d3d5309511ec2a8db)
* [v2 4/4] [UniquePersonList: Remove unnecessary method call](https://github.com/se-edu/addressbook-level4/pull/552/commits/dcbd39605225a910fc3cfbe77d1a519d9c383d5a)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/6422e8f3048fc63e9e46a8d7937eb3bc433e74ad...dcbd39605225a910fc3cfbe77d1a519d9c383d5a)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/4663d0c37ada2a3bba1aed3a766a5bd0edd9d0a1~1...4663d0c37ada2a3bba1aed3a766a5bd0edd9d0a1))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/552/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
Looks way better now. @damithc would you like to take a look at the architecture? Also, this is a new topic (not covered in syllabus atm because there were no bindings in the previous code base), so you may be interested in this.

**Comment by damithc** [comment]
I'll have to leave this decision to you @yamgent and @pyokagan because I have not much expertise in the GUI area.
I have no objections to this. The library does not bring in a new SE concept (it's more of an implementation level convenience), so it's fine to use if the benefits are worth the cost, irrespective of the syllabus.

**Comment by CanIHasReview** [comment]
# v3

@Zhiyuan-Amos submitted v3 for review.

* [v3 1/5] [Teach Person and ReadOnlyPerson to be observable](https://github.com/se-edu/addressbook-level4/pull/552/commits/d45c1605c93e4ef0a72d16ab98fdc87b900554a1)
* [v3 2/5] [Import EasyBind](https://github.com/se-edu/addressbook-level4/pull/552/commits/c7b0b910b3d25230e25638a5d07786a929d82e99)
* [v3 3/5] [PersonListPanel: Update ListView<ReadOnlyPerson> to ListView<PersonCard>](https://github.com/se-edu/addressbook-level4/pull/552/commits/5b97347626f5281aa9ebb08a371f8774075ff975)
* [v3 4/5] [UniquePersonList: Remove unnecessary method call](https://github.com/se-edu/addressbook-level4/pull/552/commits/c0cd7c9019fbdba2eba62cec827cc030221119e6)
* [v3 5/5] [PersonListPanelHandle: Update methods](https://github.com/se-edu/addressbook-level4/pull/552/commits/c830549a017a52fe1c3b37cebd3583c3f244acd4)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/56e874d0e7b4cab2b87d573782df3b34f9bafc3c...c830549a017a52fe1c3b37cebd3583c3f244acd4)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/155f2211168d1b6e0bcac4c8dfe09a8d331674e9~1...155f2211168d1b6e0bcac4c8dfe09a8d331674e9))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/552/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
@damithc for your review, Wang Leng has approved of both the architecture and the code. I'm not sure whether Paul will review or not :<

File: src/main/java/seedu/address/model/person/Person.java [hunk 2]
```diff
@@ -16,23 +18,24 @@
...
-    private Phone phone;
-    private Email email;
-    private Address address;
+    private ObjectProperty<Name> name;
+    private ObjectProperty<Phone> phone;
+    private ObjectProperty<Email> email;
+    private ObjectProperty<Address> address;
 
-    private UniqueTagList tags;
+    private ObjectProperty<UniqueTagList> tags;
```
Comment (damithc) [inline src/main/java/seedu/address/model/person/Person.java]: Are we ok with using javax/beans stuff in the model? Feels like coupling the model to the GUI

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/model/person/Person.java]: > With JavaFX it's common to use Properties for all fields of a model class.

From the guide that's acknowledged in our codebase :P http://code.makery.ch/library/javafx-8-tutorial/part2/

File: src/test/java/guitests/guihandles/PersonListPanelHandle.java [hunk 3]
```diff
@@ -61,52 +58,48 @@ public boolean isListMatching(ReadOnlyPerson... persons) throws PersonNotFoundEx
...
-                .map(PersonCardHandle::new)
-                .filter(handle -> handle.isSamePerson(person))
-                .findFirst();
-        return personCardNode.orElseThrow(PersonNotFoundException::new);
-    }
-
-    private Set<Node> getAllCardNodes() {
-        return guiRobot.lookup(CARD_PANE_ID).queryAll();
+        Optional<PersonCardHandle> handle = getRootNode().getItems().stream().filter(card -> card.person.equals(person))
+                .map(card -> new PersonCardHandle(card.getRoot())).findFirst();
```
Comment (damithc) [inline src/test/java/guitests/guihandles/PersonListPanelHandle.java]: one per line? At least line breaks at each logical step?
```java
getRootNode().getItems().stream()
        .filter(card -> card.person.equals(person))
        .map(card -> new PersonCardHandle(card.getRoot()))
        .findFirst();
```

File: src/main/java/seedu/address/model/person/Person.java (hunk 2 above)
Comment (yamgent) [inline src/main/java/seedu/address/model/person/Person.java]: I am under the impression that javax/beans is not a GUI library, even though it is commonly used for GUI logic purposes...

File: build.gradle [hunk 4]
```diff
@@ -60,6 +60,7 @@ allprojects {
     }
 
     dependencies {
+        compile group: 'org.fxmisc.easybind', name: 'easybind', version: '1.0.3'
```
Comment (damithc) [inline build.gradle]: Isn't there some other place (in dev docs) where we describe the libraries we use?

File: src/main/java/seedu/address/ui/PersonListPanel.java [hunk 5]
```diff
@@ -32,7 +34,9 @@ public PersonListPanel(ObservableList<ReadOnlyPerson> personList) {
     }
 
     private void setConnections(ObservableList<ReadOnlyPerson> personList) {
-        personListView.setItems(personList);
+        ObservableList<PersonCard> mappedList = EasyBind.map(personList, (person) ->
+                new PersonCard(person, personList.indexOf(person) + 1));
```
Comment (damithc) [inline src/main/java/seedu/address/ui/PersonListPanel.java]: This is more readable?
```java
ObservableList<PersonCard> mappedList = EasyBind.map(
        personList, (person) -> new PersonCard(person, personList.indexOf(person) + 1));
```

File: build.gradle (hunk 4 above)
Comment (Zhiyuan-Amos) [inline build.gradle]: @yamgent and I don't find any.

**Comment by CanIHasReview** [comment]
# v4

@Zhiyuan-Amos submitted v4 for review.

* [v4 1/5] [Teach Person and ReadOnlyPerson to be observable](https://github.com/se-edu/addressbook-level4/pull/552/commits/d45c1605c93e4ef0a72d16ab98fdc87b900554a1)
* [v4 2/5] [Import EasyBind](https://github.com/se-edu/addressbook-level4/pull/552/commits/c7b0b910b3d25230e25638a5d07786a929d82e99)
* [v4 3/5] [PersonListPanel: Update ListView<ReadOnlyPerson> to ListView<PersonCard>](https://github.com/se-edu/addressbook-level4/pull/552/commits/f3c9603b1d76875e4baa9ef99b3c7447f5af3739)
* [v4 4/5] [UniquePersonList: Remove unnecessary method call](https://github.com/se-edu/addressbook-level4/pull/552/commits/48dddc06f9fa2c14bfd4096d2ff2205029bf4826)
* [v4 5/5] [PersonListPanelHandle: Update methods](https://github.com/se-edu/addressbook-level4/pull/552/commits/80aa1e7650a4ccaa97d5b9d96296b4b170e0e019)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/56e874d0e7b4cab2b87d573782df3b34f9bafc3c...80aa1e7650a4ccaa97d5b9d96296b4b170e0e019)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/18df811c674ea8bf97f1a2a1cf612fbe19cc8503~1...18df811c674ea8bf97f1a2a1cf612fbe19cc8503))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/552/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by yamgent** [comment]
Propose a merge commit message.

**Comment by CanIHasReview** [comment]
# v5

@Zhiyuan-Amos submitted v5 for review.

* [v5 1/5] [Teach Person and ReadOnlyPerson to be observable](https://github.com/se-edu/addressbook-level4/pull/552/commits/9cfb265931734a8fee1a995ac22ccb12b2d3a5bb)
* [v5 2/5] [Import EasyBind](https://github.com/se-edu/addressbook-level4/pull/552/commits/ab776e0115a74f3bfd4348772c38e7aa5f04a1b5)
* [v5 3/5] [PersonListPanel: Update ListView<ReadOnlyPerson> to ListView<PersonCard>](https://github.com/se-edu/addressbook-level4/pull/552/commits/b9628364d9bd4ad0536bee28e5396d647cd4fe59)
* [v5 4/5] [UniquePersonList: Remove unnecessary method call](https://github.com/se-edu/addressbook-level4/pull/552/commits/51ed2e8d00d7de4e2b5d020ee94c8bc631674f1f)
* [v5 5/5] [PersonListPanelHandle: Update methods](https://github.com/se-edu/addressbook-level4/pull/552/commits/6a28f26f618b76e86b60a3551a169f48c4b16021)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/52ec52609b8a17da202dd3fa8f9813b7820b5d3f...6a28f26f618b76e86b60a3551a169f48c4b16021)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/0a515f876b39bb5558dcd106c9724241eae9185a~1...0a515f876b39bb5558dcd106c9724241eae9185a))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/552/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
Wrote a merge commit message.

I improved the equals method for a more thorough check, and wrote tests for it.

**Comment by yamgent** [comment]
New implementation looks okay to me. :+1:

**Comment by yamgent** [comment]
@Zhiyuan-Amos modifying the commit messages?

**Comment by Zhiyuan-Amos** [comment]
hang on, still looking through your PR :P

**Comment by CanIHasReview** [comment]
# v6

@Zhiyuan-Amos submitted v6 for review.

* [v6 1/5] [Teach Person and ReadOnlyPerson to be observable](https://github.com/se-edu/addressbook-level4/pull/552/commits/b1c5e49fad734f8b07c87178c3b3108fc6c8ee50)
* [v6 2/5] [Import EasyBind](https://github.com/se-edu/addressbook-level4/pull/552/commits/23fd697b2f4ea2de1054cdee6bda7050d26891c9)
* [v6 3/5] [PersonListPanel: Update ListView<ReadOnlyPerson> to ListView<PersonCard>](https://github.com/se-edu/addressbook-level4/pull/552/commits/75cbfa3804ff365646e903514282d9313da4300e)
* [v6 4/5] [UniquePersonList: Remove unnecessary method call](https://github.com/se-edu/addressbook-level4/pull/552/commits/6f8c65369437191e86cb3484308dc39a48e47dd8)
* [v6 5/5] [PersonListPanelHandle: Update methods](https://github.com/se-edu/addressbook-level4/pull/552/commits/baa5988d5bcf9c0d95d3aca56e90ef19e42206e8)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/84d6516fb88cceba44c55050d8cbd2a905f53374...baa5988d5bcf9c0d95d3aca56e90ef19e42206e8)) ([:chart_with_upwards_trend: Interdiff between v5 and v6](https://github.com/se-edu/addressbook-level4/compare/e180a6683af45a570b33d745c6b488683b8c8088~1...e180a6683af45a570b33d745c6b488683b8c8088))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/552/6/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
Done! @yamgent can merge?

**Pull request merged**